### ISSUE TITLE
Issue categories count

### ISSUE THREAD

**Issue by miketaylr** [webcompat/webcompat.com#291]
As part of #266, we want to have the number of issues in each category. I think this is a cool feature, but will require a creative solution. In order to know this, we could do the following:

1) traverse all issues, categorizing and counting as we go (too expensive, API-wise)
2) ask the API for each of the categories (closed and the ones with labels are the easiest, we can deduce New/Untriaged from what's remaining) and count the length. We can get `open_issues_count` from https://developer.github.com/v3/repos/#get.

I think 2 is probably the best. That's probably 5 API requests--so it might be good to store the values in a database and update it at some interval. Any advice @karlcow?

**Comment by karlcow** [comment]
Not usable directly in Flask but you got the (hackish) spirit of it.

```
→ python github-webc.py 
Today: 2016-02-24T15:41:53.706386
464 open issues
----------------------
needsinfo       12
needsdiagnosis  110
needscontact    103
contactready    80
sitewait        101
----------------------
```
https://github.com/karlcow/webcompat/blob/master/moz/minutes/github-webc.py

**Comment by miketaylr** [comment]
Depends on #886.

**Comment by karlcow** [comment]
Yes Milestones give us that for free.

**Comment by karlcow** [comment]
OK Let's revisit this one. :)

Some possibilities. Total number.

<img width="438" alt="capture d ecran 2017-11-01 a 14 20 12" src="https://user-images.githubusercontent.com/505230/32261163-1d32bf7e-bf10-11e7-85fc-dbd42e192005.png">


needstriage

<img width="437" alt="capture d ecran 2017-11-01 a 14 21 01" src="https://user-images.githubusercontent.com/505230/32261167-26692cea-bf10-11e7-80f8-b3fb538e01a3.png">

or 

<img width="426" alt="capture d ecran 2017-11-01 a 14 34 20" src="https://user-images.githubusercontent.com/505230/32261430-ce7a8e1e-bf11-11e7-95e3-26b6f52c6d2b.png">


and… obviously with a styling that would make sense. 😛

**Comment by karlcow** [comment]
to note that the markup already exists for it.

```html
<button type="button" class="wc-Tag wc-Tag--filter wc-Tag--needstriage r-ResetButton js-Tag is-active" data-filter="needstriage">
      <span class="wc-Tag-count"></span> Needs Triage
    </button>
```

https://github.com/webcompat/webcompat.com/blob/93662e358169e7443445217fe7d2a5c6151c9b21/webcompat/templates/web_modules/tag.html#L5-L11

**Comment by miketaylr** [comment]
@marimeireles this might be an interesting bug for you to work on, when you need a break from performance hacking. :)

**Comment by marimeireles** [comment]
Yea :)
Thanks, I'll keep this in mind

**Comment by laghee** [comment]
@marimeireles Once we get the database up and running for metrics (coming soon to a server near you!), all the counts will be stored there and updated hourly.

**Comment by karlcow** [comment]
@marimeireles do you still intend to work on it?
If not I will steal it ;)

Without a DB, we can already do it. 
https://api.github.com/repos/webcompat/web-bugs/milestones

This returns a JSON, for each milestones you get also the number of `closed_issues` and `open_issues`, let's say `json_data`.

Then this JSON could be reduced to this dictionary in python.

```python 
{u'worksforme': 9920, u'invalid': 6218, u'needscontact': 274, u'sitewait': 645, u'needstriage': 210, u'duplicate': 5451, u'needsdiagnosis': 168, u'contactready': 94, u'wontfix': 276, u'fixed': 1173, u'incomplete': 3100}
```

This can be done with:

```python
def get_milestones_stats(json_data):
    """Returns a count of values in all milestones."""
    stats = {}
    open_milestones = ['needstriage', 'needsdiagnosis', 'needscontact',
                        'contactready', 'sitewait']
    closed_milestones = ['duplicate', 'wontfix', 'worksforme', 'fixed',
                            'noncompat', 'invalid', 'incomplete', ]
    data = json.loads(json_data)
    for milestone in data:
        if milestone['title'] in open_milestones:
            stats[milestone['title']] = milestone['open_issues']
        if milestone['title'] in closed_milestones:
            stats[milestone['title']] = milestone['closed_issues']
    return stats
```

To think. Caching to avoid to make a request every time the page is loaded.

**Comment by marimeireles** [comment]
Sorry for the delay @karlcow. I had forgot this issue.

>https://api.github.com/repos/webcompat/web-bugs/milestones

Nice found!

No, I can tackle this if you don't mind! I don't know much about caching and I'd like to learn. I don't think we can use our current system, since it just treats file caches.

So, what do you think we should do regarding cache here? 
- A map on the backend with a simple refresh control? Like a function that refreshes the numbers every 24h for instance;
- Or maybe... Does Flask has any support for that, like built-in functions, I just found this [extension](https://pythonhosted.org/Flask-Cache/);
- I was also reading about [varnish](https://varnish-cache.org/intro/index.html#intro) and even though it's definitely an overkill resolution for this bug, it may be useful for your future server side rendering plans.

Hmm, also, as always I'm not sure about the best place to put this `get_milestones_stats` function. You think `webcompat/helpers.py` is a good place?

**Comment by karlcow** [comment]
* The function for stats would land in helpers, yes. 
* It requires another function for fetching the milestones json aka `json_data` returned by https://api.github.com/repos/webcompat/web-bugs/milestones and which make sure that this would work with webcompat-tests too. 

As for caching and not do the request every time. Let's see if we can trust the browser.

First request to https://api.github.com/repos/webcompat/webcompat-tests/milestones

```http
HTTP/1.1 200 OK
Server: GitHub.com
Date: Fri, 17 May 2019 01:27:39 GMT
Content-Type: application/json; charset=utf-8
Transfer-Encoding: chunked
Status: 200 OK
X-RateLimit-Limit: 60
X-RateLimit-Remaining: 58
X-RateLimit-Reset: 1558060059
Cache-Control: public, max-age=60, s-maxage=60
Vary: Accept
ETag: W/"4a841db6685f2700eebe34e7d9763bf1"
X-GitHub-Media-Type: unknown, github.v3
Access-Control-Expose-Headers: ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type
Access-Control-Allow-Origin: *
Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
X-Frame-Options: deny
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
Referrer-Policy: origin-when-cross-origin, strict-origin-when-cross-origin
Content-Security-Policy: default-src 'none'
Content-Encoding: gzip
X-GitHub-Request-Id: EBCC:4331:7E2405:140693A:5CDE0E0B
```

Reload

```http
HTTP/1.1 304 Not Modified
Date: Fri, 17 May 2019 01:32:55 GMT
Content-Type: application/octet-stream
Server: GitHub.com
Status: 304 Not Modified
X-RateLimit-Limit: 60
X-RateLimit-Remaining: 58
X-RateLimit-Reset: 1558060375
Cache-Control: public, max-age=60, s-maxage=60
ETag: "4a841db6685f2700eebe34e7d9763bf1"
Access-Control-Expose-Headers: ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type
Access-Control-Allow-Origin: *
Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
X-Frame-Options: deny
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
Referrer-Policy: origin-when-cross-origin, strict-origin-when-cross-origin
Content-Security-Policy: default-src 'none'
Vary: Accept, Accept-Encoding
X-GitHub-Request-Id: FF88:5A27:1A55483:2022F63:5CDE0F47
```

Let's modify an issue to change the milestone, hence the count in each milestone.

```http
HTTP/1.1 200 OK
Date: Fri, 17 May 2019 01:35:23 GMT
Content-Type: application/json; charset=utf-8
Transfer-Encoding: chunked
Server: GitHub.com
Status: 200 OK
X-RateLimit-Limit: 60
X-RateLimit-Remaining: 57
X-RateLimit-Reset: 1558060523
Cache-Control: public, max-age=60, s-maxage=60
ETag: W/"53f0d828f9c2a4c47d84a4d813f33f09"
X-GitHub-Media-Type: unknown, github.v3
Access-Control-Expose-Headers: ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type
Access-Control-Allow-Origin: *
Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
X-Frame-Options: deny
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
Referrer-Policy: origin-when-cross-origin, strict-origin-when-cross-origin
Content-Security-Policy: default-src 'none'
Content-Encoding: gzip
Vary: Accept, Accept-Encoding
X-GitHub-Request-Id: C8CF:27BC:1D55BF3:23D31CC:5CDE0FDB
```

<table>
<tr>
  <th>sequence</th>
  <th>Response code</th>
  <th>Cache-Control</th>
  <th>ETAG</th>
</tr>
<tr>
  <td>1st</td>
  <td>200 OK</td>
  <td>public, max-age=60, s-maxage=60</td>
  <td>W/"4a841db6685f2700eebe34e7d9763bf1"</td>
</tr>
<tr>
  <td>1st (a couple of minutes later)</td>
  <td>304 Not Modified</td>
  <td>public, max-age=60, s-maxage=60</td>
  <td>"4a841db6685f2700eebe34e7d9763bf1"</td>
</tr>
<tr>
  <td>3rd (changing a milestone)</td>
  <td>200 OK</td>
  <td>public, max-age=60, s-maxage=60</td>
  <td>W/"53f0d828f9c2a4c47d84a4d813f33f09"</td>
</tr>
</table>

So the system works well, even if we reload a page it will not fetch again the milestone file, but just use the browser's cache. Cool.

**Comment by karlcow** [comment]
@marimeireles so I guess no need to worry about the caching as the browser is already doing its job.

**Comment by marimeireles** [comment]
Okay then.
Thanks for testing!

**Comment by karlcow** [comment]
I propose we close this. Nobody is willing to work on it and maybe it's not a priority.
